### ISSUE TITLE
[PHP] Move to a modern output

### ISSUE THREAD

**Issue by ferdynator** [swig/swig#1529]
PHP has come great lengths in the last years. The currently generated swig output does not meet best-practices, uses [deprecated functions](https://php.net/manual/en/function.dl.php)
Namespaces are not supported, [type-hinting](https://www.php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration) and [function return types](https://www.php.net/manual/en/functions.returning-values.php). Since swig dropped support for older PHP versions and the default PHP version is `7` many things could be improved.

I would like to implement (or help implementing) some of these features but don't have the most experience with swig. I've already added a PR #1528 to remove a closing tag, because that is considered bad practice. What would be the best first / next steps to take here?

Some languages (as C#) have `namespace` support which is a feature I would love to implement for PHP. Modern PHP applications use a JAVA-style _1-class-per-file_ setup. Does it make sense to copy from the [java implementation](https://github.com/swig/swig/blob/master/Source/Modules/java.cxx)? There are [some FIXMEs](https://github.com/swig/swig/blob/master/Source/Modules/php.cxx#L15) in the php code. Should they be addressed first? 

It seems like PHP is rarely used and thus the generation is quite out of date too. But with upcoming [JIT](https://wiki.php.net/rfc/jit) PHP is getting more interesting for CPU heavy load. And having swig generate nice PHP modules out of the box would be awesome :)

**Comment by ojwb** [comment]
The answer to most of these is just that nobody's implemented it yet.

For namespace support there's already a ticket open: #69

The use of `dl()` is just there as a fallback to try if the C glue module isn't already loaded.  I've wondered about removing it entirely, but worried it would break things for some people running code with CLI PHP.  I guess if we're going to, now's a fairly good time (doing it for 4.0.0 would have been better though).

I don't think moving to one PHP file per class is helpful though - it makes sense for code humans need to maintain, but for generated code it doesn't really help, and having everything in one file is actually more convenient to deal with.  And getting the work from the Nihal's GSoC project merged would change us to using the PHP C API to create classes instead of generating PHP code to define classes which wrap flat functions, and that would make this point moot.

**Comment by ferdynator** [comment]
Are you talking about #1092 ? What can be done to get this merged? I'll be looking into namespaces later this week. Also I can open a PR for the removal of `dl()`. You can merge it any time it fits then.

**Comment by ojwb** [comment]
https://github.com/swig/swig/tree/gsoc2017-php7-classes-via-c-api is the branch.

**Comment by ojwb** [comment]
@ferdynator Did you look into removing use of `dl()`?

We're working towards a SWIG 4.2.0 release, which would be a good time to stop generating the code using `dl()`, I had a look at this, but got stuck at making SWIG's testsuite work without it.  For reasons I don't understand, some tests run fine, some segfault, some others complain about missing globals (each testcase checks the list of globals against those it expects the testcase to create).

**Comment by ferdynator** [comment]
It has been some time. But I would love to move this forward. For `dl` there is no real replacement other than loading the extension via php.ini files. I can checkout the testcases

**Comment by ojwb** [comment]
Let's reopen this, as most of the issues raised are still valid.

I'm lacking in SWIG-hacking time currently, so assistance with improving the PHP backend is definitely appreciated and I'll try to answer questions and review patches promptly.

I'm trying to run testcases like this, which is effectively specifying `php.ini` via command line options:

```
$(PHP) -n -q -d extension_dir=. -d display_errors=stderr -d extension=$(PHP_EXTENSION)$(PHP_SO) $(PHP_SCRIPT)
```

We could generate a temporary `php.ini` instead, though it's more awkward to manage if testcases can be run in parallel (not actually sure if we support that or not.)

I think we can actually drop `-q` which  doesn't seem to do anything for current CLI PHP.  The SWIG git history isn't helpful, but looking at xapian-bindings git history I found:

> pass the "-q" option, which suppresses HTTP header output for PHP < 4.3 and is ignored for compatibility by more recent versions of PHP.

I'll push a WIP branch for my work on `dl()` elimination so far.

**Comment by ojwb** [comment]
I've just tried generating a `php.ini` instead of using `-d` and that doesn't seem to make any difference (testing with PHP 8.0):

```
index 3372b52c..84a43593 100644
--- a/Examples/Makefile.in
+++ b/Examples/Makefile.in
@@ -1077,7 +1077,8 @@ php_cpp: $(SRCDIR_SRCS)
 # -----------------------------------------------------------------
 
 php_run:
-       $(RUNTOOL) $(PHP) -n -d extension_dir=. -d extension=$(PHP_EXTENSION) -d display_errors=stderr $(PHP_SCRIPT) $(RUNPIPE)
+       printf '[PHP]\nextension_dir=./\nextension=%s\ndisplay_errors=stderr\n' '$(PHP_EXTENSION)' > php.ini
+       $(RUNTOOL) $(PHP) -n --php-ini php.ini $(PHP_SCRIPT) $(RUNPIPE)
 
 # -----------------------------------------------------------------
 # Version display
```

And a backtrace from gdb suggested this as related to the code that changed in 71475b0af9677deeaf6fe55c0c5f53fec9f730d2, but reverting to what we had before doesn't seem to help, and neither does changing to the other alternative suggested in the ticket.

I think I'm out of ideas.

One thing to beware of if anyone is hacking on this is that changing `Examples/Makefile.in` and running `make` or `make check-php-test-suite` doesn't seem to regenerate `Examples/Makefile` automatically - you need to run `./config.status Examples/Makefile` from the top of the source tree by hand for changes to actually take effect.  There is a rule that's meant to take care of this near the top of `Examples/Makefile.in` but it seems nothing triggers it in these cases.

**Comment by ferdynator** [comment]
So I was able to run an example with the following:

```
diff --git a/Examples/Makefile.in b/Examples/Makefile.in
index 3372b52c2..8752a0112 100644
--- a/Examples/Makefile.in
+++ b/Examples/Makefile.in
@@ -1077,7 +1077,7 @@ php_cpp: $(SRCDIR_SRCS)
 # -----------------------------------------------------------------
 
 php_run:
-       $(RUNTOOL) $(PHP) -n -d extension_dir=. -d extension=$(PHP_EXTENSION) -d display_errors=stderr $(PHP_SCRIPT) $(RUNPIPE)
+       $(RUNTOOL) $(PHP) -n -d extension_dir=. -d extension=example$(PHP_SO) -d display_errors=stderr $(PHP_SCRIPT) $(RUNPIPE)
```

This works on the current `php-stopusing-dl` branch.

I can spend some time working on the resulting PHP files but I probably lack the skill to work on any C/C++ internals, sorry. If you have something I should take a look at, let me know. Otherwise I'll try to move towards multi-file output for _namespaced_ code.

**Comment by ojwb** [comment]
> (The PHP_EXTENSION variable was simply holding `pwd`, I don't exactly know why but this definitely wont work)

The value set there is used for the examples (i.e. `Examples/php/*`) - for the testsuite it's overridden when `Examples/test-suite/php/Makefile` runs the `php_run` target here.  I had set it to `pwd` for the examples thinking it wanted to be the name of the example, but without really thinking it through (firstly `pwd` has the path on, and secondly the shared object is actually named `example` rather than having the name of the example unlike the testsuite where it is named after the testcase) and I hadn't actually testing with the examples - it should actually be `PHP_EXTENSION = example$(PHP_SO)`, like you have but specified via `PHP_EXTENSION` so the testsuite can override.  I'll apply that change.

But with your change I get segfaults with one example, so this doesn't actually solve the "some things work, others segfault" that I saw with the testsuite:

```
$ make check-php-examples
checking Examples/php/callback
checking Examples/php/class
checking Examples/php/constants
checking Examples/php/cpointer
checking Examples/php/disown
checking Examples/php/enum
checking Examples/php/extend
checking Examples/php/funcptr
Segmentation fault
make[3]: *** [../../Makefile:1080: php_run] Error 139
make[2]: *** [Makefile:11: check] Error 2
make[1]: *** [Makefile:216: funcptr.actionexample] Error 2
checking Examples/php/overloading
checking Examples/php/pointer
checking Examples/php/pragmas
checking Examples/php/proxy
checking Examples/php/reference
checking Examples/php/simple
checking Examples/php/sync
checking Examples/php/value
checking Examples/php/variables
make: *** [Makefile:201: check-php-examples] Error 2
```

I'd suggest not putting a lot of work into changing the generated `.php` as I'd like to get the gsoc2017-php7-classes-via-c-api branch merged and then we'd be creating everything via PHP's C API - there'd just be the shared object and no `.php` files at all.

**Comment by ojwb** [comment]
After some more poking at around it looks to me like the issue is with setting up a PHP constant which is a PHP resource (which we do for pointers, references and arrays we don't have full type information for).  For some reason this seems to work OK when the extension is loaded with `dl()`, but not when it's loaded from `php.ini`.  I suspect this also means that this feature can't currently be used except with the CLI PHP - I don't recall any reports of this, perhaps as wrapping such things is a bit esoteric and not actually widely used in production.

The internals docs try to steer the reader away from using resources and towards using objects instead, but that doesn't help here as the same docs say you can't create a PHP constant which is a PHP object or array (but it doesn't include resources in that list.)

**Comment by ferdynator** [comment]
I was able to reproduce the issues on php 7.4. How do I compile the examples with debug symbols so I can look into the problem with gdb?

**Comment by ojwb** [comment]
@ferdynator for example e.g.:

```
cd Examples/php/funcptr
make CFLAGS='-g' CXXFLAGS='-g' RUNTOOL='gdb --args'
```

And for testcases:

```
cd Examples/test-suite/php
make CFLAGS='-g' CXXFLAGS='-g' RUNTOOL='gdb --args' callback.test
```

These just give you debug symbols for the code in the extension module - you might also want to install the debug symbols package(s) for PHP itself, or rebuild it with debug symbols.

**Comment by ojwb** [comment]
Sorry, I noticed I failed to respond to this before:

> There are some FIXMEs in the php code. Should they be addressed first?

Reviewing every FIXME and TODO in `php.cxx`:

```
 * Medium term:
 *
 * Handle default parameters on overloaded methods in PHP where possible.
 *   (Mostly done - just need to handle cases of overloaded methods with
 *   default parameters...)
 *   This is an optimisation - we could handle this case using a PHP
 *   default value, but currently we treat it as we would for a default
 *   value which is a compound C++ expression (i.e. as if we had a
 *   method with two overloaded forms instead of a single method with
 *   a default parameter value).
```

This one will become irrelevant if we move to creating classes using the C API as the whole messy generated PHP class wrapper layer will go away.

```
 * Sort out locale-dependent behaviour of strtod() - it's harmless unless
 *   SWIG ever sets the locale and DOH/base.c calls atof, so we're probably
 *   OK currently at least.
```

Nothing else in SWIG seems to worry about locale-dependent floating point parsing and formatting, so I guess it's pointless to worry about it here - if we wanted to set the locale in the swig tool itself there would be a lot of work to do.  I think I'll just remove this and related notes.

```
/*
 * TODO: Replace remaining stderr messages with Swig_error or Swig_warning
 * (may need to add more WARN_PHP_xxx codes...)
 */
```

Only 2 stderr message cases seem to remain - `malloc()` failing in two places (see below), and what's essentially an assertion with hooking the type tracing function.  These aren't appropriate for `Swig_error`/`Swig_warning` which report a problem relating to a file (with optional line number), so this TODO has actually already been addressed.

```
 * FIXME: Isn't using a leading underscore a bit suspect here?
 */
#define SWIG_PTR "_cPtr"
```

Not sure if it's suspect or not to have a PHP property name start with an underscore, but I think this special property will go away if we create classes via the API.

```
        /* FIXME: How should this be handled?  The rest of SWIG just seems
         * to not bother checking for malloc failing! */
        fprintf(stderr, "Malloc failed!\n");
        SWIG_exit(EXIT_FAILURE);
```

I wrote that comment when I was pretty new to the codebase.

Sadly it's still pretty common for swig to segfault if `malloc()` returns `NULL`, but places that do check seem to do pretty much the above.  I think the best solution would be to have a `safe_malloc()` or similar wrapper for use in the swig tool itself which dealt with the error case and could be relied on to return a non-NULL pointer (or to not return), but I don't really have the time to get side-tracked into that.

For now I'll just remove these notes from the PHP backend.

```
            // This handling is safe, but I'm wondering if it may be overly     
            // conservative (FIXME) in some cases.  It seems it's only bad when                                                                                    
            // there's an overloaded form with the appropriate number of
            // parameters which doesn't want the default value, but I need to                                                                                                   
            // think about this more.                                     
```

I think this will become irrelevant if we eliminate the PHP class wrapper layer.

```
          // FIXME: (really_overloaded && handle_as_overload) is perhaps a                                                                                         
          // little conservative, but it doesn't hit any cases that it    
          // shouldn't for Xapian at least (and we need it to handle                                                                                                            
          // "Enquire::get_mset()" correctly).                            
          bool non_php_default = ((really_overloaded && handle_as_overload) ||     
                                  !value || strcmp(value, "?") == 0);                                       
```

Ditto.

```
              /*                                                               
               * _p_Foo -> Foo, _p_ns__Bar -> Bar                               
               * TODO: do this in a more elegant way                                                                                                               
               */                                                         
```

Ditto.

```
            // FIXME: enhance to handle `<previous_enum> + 1' which is what                                                                                        
            // we get for enums that don't have an explicit value set.    
```

Ditto.

```
        Printf(s_phpclasses, "\n\tfunction __set($var,$value) {\n");            
        // FIXME: tune this threshold...                                                                                                                           
        if (Len(shadow_set_vars) <= 2) {                                  
          // Not many setters, so avoid call_user_func.                                                                                                                         
```

Ditto.  (Maybe it's starting to become clear why I'm keen to get rid of this layer!)

```
          /* TODO check this */
          if (Getattr(n, "wrap:disown")) {
            Replaceall(tm, "$disown", "SWIG_POINTER_DISOWN");
          } else {
            Replaceall(tm, "$disown", "0");
          }
```

Not sure what this is about - I don't fully grasp the "disown" stuff.  The comment comes from 0249eea38995dfc6689c78cde861b7ec2b6b4af2 which is the merge of the implementation of directors for PHP.

This one would be good to resolve.  Probably the way forward would be to add `*runme.php` for any director-related testcases which don't have one already (based on those for other languages) and fix any issues that uncovers.  If all the director-related testcases have a run-time test and they all pass we can be pretty confident that the director implementation is right.

**Comment by ojwb** [comment]
#1982 is now passing the testsuite, and on that the `.php` wrapper now only contains this code to use `dl()` (unless the user specifies php code to inject into it, but going forwards we can drop that feature and people can just ship a `.php` file to wrap the compiled extension if they have hand-written PHP code to augment the SWIG-generated extension. This PR addresses most of the FIXME/TODO comments in php.cxx.

I tried merging the `php-stopusing-dl` branch into the #1982 branch, but it didn't work.  I've not investigated why yet.

**Comment by ojwb** [comment]
>           /* TODO check this */
>           if (Getattr(n, "wrap:disown")) {
>             Replaceall(tm, "$disown", "SWIG_POINTER_DISOWN");
>           } else {
>             Replaceall(tm, "$disown", "0");
>           }
> ```
> 
> Not sure what this is about - I don't fully grasp the "disown" stuff. The comment comes from [0249eea](https://github.com/swig/swig/commit/0249eea38995dfc6689c78cde861b7ec2b6b4af2) which is the merge of the implementation of directors for PHP.

I just noticed that this comment was actually copied from the director implementation in `python.cxx`, but I don't really know what wants checking there either but if `wrap:disown` gets exercised with directors by the testsuite then this can probably be removed.

**Comment by ojwb** [comment]
Looking at the history it was present in the original commit of director support for Python back in 2003 (c47494cb7a725bc3355fc1c89d220935f98b02c7).

**Comment by ojwb** [comment]
#1982 is now green for CI (though there's at least one issue if you run the testsuite building with ubsan), but some testcases still crash with these changes added on top.

Simplest way to get them seems to be:

```
git cherry-pick 338cdd068530e653a7200581cb834a74f71d457f 393e91a2510ac52b4d4178f3e81c1cd0caba92ff
```

This just needs a conflict on the PHP command line in `Examples/Makefile.in` to be fixed.

**Comment by ojwb** [comment]
I've got this working on the branch in #1982.

The main underlying problem was that certain things crash if attempted in the MINIT if the module is loaded via `extension=` because the zend runtime hasn't been initialised enough to support them.  I've delayed the ones which crashed in the testsuite (which were creating objects and looking up classes (so we can add them as interfaces to a class) until RINIT time.  This means they get redone for every request, but we're still much, much better off than with the generated wrapper file getting parsed on every request.

**Comment by ojwb** [comment]
I've merged #1982 and we're no longer generating the code which uses `dl()` on git master.

It looks like the remaining things suggested above which don't have a separate open ticket already are type hints and function return types.  (The suggestion to have one `.php` file per class is now moot, as the classes are now entirely created by C/C++ code and we no longer generate a `.php` file at all by default.)

**Comment by ojwb** [comment]
Oh, there's also adding more `_runme.php` files for the testsuite.

**Comment by ojwb** [comment]
I think it's time to close this ticket.  This have improved a lot in the last ~2.5 years and reviewing, everything I can see discussed here is either addressed or has its own ticket now, except for "adding more `_runme.php` files for the testsuite".

If someone has time and energy to work on that it would be very useful to do, and would almost certainly uncover a few bugs (e.g. the regression test in #2112 just implements the previously missing `Examples/test-suite/php/newobject2_runme.php`), but that's probably not worth a ticket.  I don't think we have a complete set of "runme" files for any SWIG target language but we don't have open tickets for any I'm aware of.

Similarly, help with the other open PHP tickets (or other useful PHP modernisation work) would be appreciated.  I can try to give pointers on the SWIG side if needed.
